### PR TITLE
Rework startup threading

### PR DESCRIPTION
**Pull request by lemnik**
## Goal
Improve the threading during startup and reduce the overheads.

## Design
Replaced the various layers of startup tasks with a dedicated structure. The `RunnableProvider` class replaces the previously used `Future`, `SafeFuture` and `Callable` / `Runnable` tasks. The `RunnableProvider` safely computes values on background threads, and blocks only when required while also avoiding running the work (such as IO) on the main thread.

## Testing
Updated tests to handle the new structures, and relied on existing tests along with manual testing with `StrictMode` using `detectAll()`.

### CODE REVIEW

File: bugsnag-android-core/src/main/java/com/bugsnag/android/StorageModule.kt [hunk 1]
```diff
@@ -2,56 +2,72 @@ package com.bugsnag.android
...
-    private fun lastRunInfo(): LastRunInfo? {
+    val lastRunInfo = lastRunInfoStore.map { lastRunInfoStore ->
         val info = lastRunInfoStore.load()
         val currentRunInfo = LastRunInfo(0, crashed = false, crashedDuringLaunch = false)
         lastRunInfoStore.persist(currentRunInfo)
-        return info
+        return@map info
+    }
+
+    fun loadUser(initialUser: User): Provider<UserState> = provider {
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/StorageModule.kt:68]: This appears to be doubly-wrapped in providers - what's the reason for UserStore being handled differently?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.kt [hunk 2]
```diff
@@ -13,14 +12,14 @@ import java.util.Date
...
+    maxPersistedSessions: Int,
+    private val apiKey: String,
     logger: Logger,
     delegate: Delegate?
 ) : FileStore(
-    File(
-        config.persistenceDirectory.value, "bugsnag/sessions"
-    ),
-    config.maxPersistedSessions,
+    File(bugsnagDir, "sessions"),
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.kt:21]: Are we consistent in when we add a `bugsnag` subdir under the persistence directory?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/internal/dag/Provider.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    override fun get(): E {
+        return getOr { throw value as Throwable }
+    }
+
+    private inline fun getOr(failureHandler: () -> E): E {
+        while (true) {
+            when (state.get()) {
+                TASK_STATE_RUNNING -> awaitResult()
+                TASK_STATE_PENDING -> {
+                    if (isMainThread()) {
```
[the remark this answers is not in the document]
Comment (lemnik) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/internal/dag/Provider.kt:57]: No. All of the `Provider` work is always carried out on background threads, so they never wait for the main thread.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/StorageModule.kt (hunk 1 above)
Comment (tomlongridge) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/StorageModule.kt:68]: Discussed IRL - other fields in client will probably go this way in future changes.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.kt (hunk 2 above)
Comment (tomlongridge) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.kt:21]: I see now - these are the uses where we go through `migrateLegacyFiles`; the others are using the config value directly.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/internal/dag/DependencyModule.kt [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+    @JvmField
+    val taskType: TaskType = TaskType.DEFAULT
+) : DependencyModule {
+    inline fun <R> provider(crossinline provider: () -> R): RunnableProvider<R> {
+        return bgTaskService.provider(taskType, provider)
+    }
+
+    internal inline fun <E, R> Provider<E>.map(crossinline mapping: (E) -> R): RunnableProvider<R> {
+        val task = object : RunnableProvider<R>() {
+            override fun invoke(): R = mapping(this@map.get())
```
Comment (YYChen01988) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/internal/dag/DependencyModule.kt:20]: Maybe nice to have a note describing what the invoke do and also this class's task?

**Review by YYChen01988 — CHANGES_REQUESTED** [review]

**Review by tomlongridge — APPROVED** [review]

**Review by YYChen01988 — APPROVED** [review]
LGTM, pending CI

**Pull request merged**